magpie/alembic/versions/2023-08-25_2cfe144538e8_add_network_tables.py

```diff
@@ -0,0 +1,54 @@
+"""
+Add Network_Tokens Table
+
+Revision ID: 2cfe144538e8
+Revises: 5e5acc33adce
+Create Date: 2023-08-25 13:36:16.930374
+"""
+import uuid
+
+import sqlalchemy as sa
+from alembic import op
+from sqlalchemy.dialects.postgresql import UUID
+from sqlalchemy.orm.session import sessionmaker
+from sqlalchemy_utils import URLType
+
+# Revision identifiers, used by Alembic.
+# pylint: disable=C0103,invalid-name  # revision control variables not uppercase
+revision = "2cfe144538e8"
+down_revision = "5e5acc33adce"
+branch_labels = None
+depends_on = None
+
+Session = sessionmaker()
+
+
+def upgrade():
+    op.create_table("network_tokens",
+                    sa.Column("token", UUID(as_uuid=True),
+                              primary_key=True, default=uuid.uuid4, unique=True),
+                    sa.Column("user_id", sa.Integer,
+                              sa.ForeignKey("users.id", onupdate="CASCADE", ondelete="CASCADE"), unique=True)
+                    )
+    op.create_table("network_nodes",
+                    sa.Column("id", sa.Integer, primary_key=True, autoincrement=True),
+                    sa.Column("name", sa.Unicode(128), nullable=False, unique=True),
+                    sa.Column("url", URLType(), nullable=False)
+                    )
+    op.add_column("users", sa.Column("network_node_id", sa.Integer,
+                                     sa.ForeignKey("network_nodes.id", onupdate="CASCADE", ondelete="CASCADE"),
+                                     nullable=True))
+    op.drop_constraint("uq_users_user_name", "users")
```

Comment:
Yeah I tried to search through to catch all the cases for the current searches and enforce that the associated network node be null. Another pair of eyes would be great.

Reply:
In general, whenever `<user>.user_name` is accessed in the code, that `<user>` reference must be resolved beforehand based on what was passed as input from the API, UI, CLI or configuration startup.

This is an important one: 
https://github.com/Ouranosinc/Magpie/blob/1d602eedbe4047fd95c4d2684b5731f4d0600977/magpie/api/requests.py#L193

There are also some considerations to be taken for 
https://github.com/Ouranosinc/Magpie/blob/1d602eedbe4047fd95c4d2684b5731f4d0600977/magpie/models.py#L210
and 
https://github.com/Ouranosinc/Magpie/blob/1d602eedbe4047fd95c4d2684b5731f4d0600977/magpie/models.py#L921

which generates a somewhat related definition to a future user, but that is only really "linked" via `user_name` until the relevant registration/approval process is completed, since the actual `user_id` is not defined yet. 

This code never checks for potential multi-results returned from `user_name` matches because it always assumed the unique constraint. 

Similarly, all webhooks events for interactions with Cowbird work using `user_name`, since this is what the API paths use.